### PR TITLE
Implement per-user rate limiting

### PR DESCRIPTION
**Pull request by Brainicism**
Closes #1025

### CODE REVIEW

File: src/ratelimiter.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
+export default class RateLimiter {
```
Comment (taahamahdi) [inline src/ratelimiter.ts:1]: rate_limiter.ts, similar change to test

File: src/test/ci/ratelimiter.test.ts [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+describe("multiple users", () => {
+    const userID2 = "234";
+    describe("user 1 hits limit, user 2 doesn't", () => {
+        it("should pass all requests until capacity for user 1, and fail else. user 2 should all pass", () => {
+            for (let i = 0; i < queueCapacity; i++) {
+                assert.strictEqual(rateLimiter.check(userID), true);
+            }
+
+            for (let i = 0; i < queueCapacity / 2; i++) {
```
Comment (taahamahdi) [inline src/test/ci/ratelimiter.test.ts:138]: Could go up to `queueCapacity - 1` to be a bit more strict with the check

File: src/ratelimiter.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    /**
+     * @param userID - the user's ID
+     * @returns the time remaining until the next request can be processed, in milliseconds
+     */
+    public timeRemaining(userID: string): number {
+        if (!(userID in this.limitMap)) return 0;
+        const previousRequestDates = this.limitMap[userID];
+        return (previousRequestDates[0] + (this.queueExpiryTime * 1000)) - Date.now();
```
Comment (taahamahdi) [inline src/ratelimiter.ts:19]: Can this ever be negative? If it can, should return 0 in that case

**Review by taahamahdi — CHANGES_REQUESTED** [review]

File: src/test/ci/ratelimiter.test.ts (hunk 2 above)
Comment (Brainicism) [reply src/test/ci/ratelimiter.test.ts:138]: This test tests that the requests from one user doesn't impact the requests from another user, intentionally did not fill it to capacity

**Review by taahamahdi — APPROVED** [review]

**Pull request merged**